agent/agent_test.go

```diff
@@ -856,6 +857,98 @@ func testAgent_AddServiceNoRemoteExec(t *testing.T, extraHCL string) {
 	}
 }
 
+func TestCacheRateLimit(test *testing.T) {
+	test.Parallel()
+	tests := []struct {
+		// count := number of updates performed (1 every 10ms)
+		count int
+		// rateLimit rate limiting of cache
+		rateLimit float64
+		// Minimum number of updates to see from a cache perspective
+		// We add a value with tolerance to work even on a loaded CI
+		minUpdates int
+	}{
+		// 250 => we have a test running for at least 2.5s
+		{250, 0.5, 1},
+		{250, 1, 1},
+		{300, 2, 2},
+	}
+	for _, currentTest := range tests {
+		test.Run(fmt.Sprintf("rate_limit_at_%v", currentTest.rateLimit), func(t *testing.T) {
+			tt := currentTest
+			t.Parallel()
+			a := NewTestAgent(t, fmt.Sprintf("cache = { rate_limit_per_entry = %v, rate_limit_burst = 1 }", tt.rateLimit))
```

Comment:
Do we need a full agent to test this change?

It seems like the change is fairly localized to the `agent/cache` package, so I would expect the test to be in that same package. If we use a fake `cache.Type` implementation the test should read better as there will be fewer things that are not related to the behaviour we are trying to test. A fake implementation could report on the number of times it was called, which should make it easy to test that refresh is blocked on a limit.

That should also make it easier to show that different keys do not share a rate limit.

Reply:
I think it is the best proof it is working from end-to-end. What would you suggest exactly?

I am not very familiar with Mocks (and to be fair, not always a big fan), but if you point me to an example of what you have in mind...